pjlib/src/pj/ssl_sock_ossl.c

```diff
@@ -455,6 +455,139 @@ static int openssl_init_count;
+{
+    return ((unsigned long)pj_thread_get_os_handle(pj_thread_this()));
+}
+#endif
+
+static void ossl_lock(int mode, int id, const char *file, int line)
+{
+    PJ_UNUSED_ARG(file);
+    PJ_UNUSED_ARG(line);
+
+    if (openssl_init_count == 0)
+        return;
+
+    if (mode & CRYPTO_LOCK) {
+        if (ossl_locks[id]) {
+            //PJ_LOG(6, (THIS_FILE, "Lock File (%s) Line(%d)", file, line));
+            pj_lock_acquire(ossl_locks[id]);
+        }
+    } else {
+        if (ossl_locks[id]) {
+            //PJ_LOG(6, (THIS_FILE, "Unlock File (%s) Line(%d)", file, line));
+            pj_lock_release(ossl_locks[id]);
+        }
+    }
+}
+
+static void release_thread_cb(void)
+{
+    unsigned i = 0;
+
+#if     OPENSSL_VERSION_NUMBER >= 0x10000000
+    CRYPTO_THREADID_set_callback(NULL);
+#else
+    CRYPTO_set_id_callback(NULL);
+#endif
+    CRYPTO_set_locking_callback(NULL);
+
+    for (; i < ossl_num_locks; ++i) {
+        if (ossl_locks[i]) {
+            pj_lock_destroy(ossl_locks[i]);
+            ossl_locks[i] = NULL;
+        }
+    }
+    if (lock_pool) {
+        pj_pool_release(lock_pool);
+        lock_pool = NULL;
+        pj_caching_pool_destroy(&cp);
+    }
+    ossl_locks = NULL;
+    ossl_num_locks = 0;
+}
+
+static pj_status_t init_ossl_lock()
+{
+    pj_status_t status = PJ_SUCCESS;
+
+    pj_caching_pool_init(&cp, NULL, 0);
+
+    lock_pool = pj_pool_create(&cp.factory,
+                               "ossl-lock",
+                               64,
+                               64,
+                               NULL);
+
+    if (!lock_pool) {
+        status = PJ_ENOMEM;
+        PJ_PERROR(1, (THIS_FILE, status,"Fail creating OpenSSL lock pool"));
+        pj_caching_pool_destroy(&cp);
+        return status;
+    }
+
+    ossl_num_locks = CRYPTO_num_locks();
+    ossl_locks = (pj_lock_t **)pj_pool_calloc(lock_pool,
+                                              ossl_num_locks,
+                                              sizeof(pj_lock_t*));
+
+    if (ossl_locks) {
+        unsigned i = 0;
+        for (; (i < ossl_num_locks) && (status == PJ_SUCCESS); ++i) {
+            status = pj_lock_create_simple_mutex(lock_pool, "ossl_lock%p",
+                                                 &ossl_locks[i]);
+        }
+        if (status != PJ_SUCCESS) {
+            PJ_PERROR(1, (THIS_FILE, status,
+                          "Fail creating mutex for OpenSSL lock"));
+            release_thread_cb();
+            return status;
+        }
+
+#if     OPENSSL_VERSION_NUMBER >= 0x10000000
+        CRYPTO_THREADID_set_callback(ossl_set_thread_id);
+#else
+        CRYPTO_set_id_callback(ossl_thread_id);
+#endif
+        CRYPTO_set_locking_callback(ossl_lock);
+        status = pj_atexit(&release_thread_cb);
+        if (status != PJ_SUCCESS) {
+            PJ_PERROR(1, (THIS_FILE, status,
+                  "Cannot set OpenSSL lock thread callback unrelease method"));
+            release_thread_cb();
```

Comment:
hmm.. not sure if we should uninstall the callbacks when atexit() fails, just print warning I guess (btw, perhaps it is good idea to explicitly put keyword 'Warning!' in the log), usually app is about to be destroyed too after the library is uninitialized?

Reply:
if we are allowing the thread lock to continue it's work without having the chance to release the resources  (e.g: pool, locks) properly, wouldn't it be considered a leak?

Reply:
yes, aware of the leak, but perhaps here are things we can consider:
- atexit failure will not disrupt the lib operations, while cancelled thread lock can be harmful (btw, please add description into PR about the background or reported problem).
- as mentioned before, usually lib destroy (where the leak happens) will be followed by app destroy soon (where OS will resolve the leak).